docs/rfcs/025-generation-numbers.md

```diff
@@ -0,0 +1,650 @@
+deletion objects with latest metadata to ensure all victim objects are really not referenced by
+latest metadata.
+
+### Deletion Part 2: WAL trim changes
+
+Remote objects are not the only kind of deletion the pageserver does: it also indirectly deletes
+WAL data, by feeding back remote_consistent_lsn to safekeepers, as a hint to the safekeepers that
+they may drop data below this LSN.
+
+We may solve for safety issues here in the same way as for deletions: before indicating to safekeepers
+that they may trim up to a particular LSN, pageservers should communicate with the control plane
+to confirm that their node generation is current. This guarantees that any future generation will
+be starting from the last metadata that this generation wrote out, i.e. the S3 state which contains
+all the data from the WAL up to the proposed trim point.
+
+The pageservers indicate what to trim by the `remote_consistent_lsn` field in messages to
+the safekeeper. The pageservers will change to only advertise the true remote_consistent_lsn
+after they have periodically checked the validity of their generation: this will slightly
+delay WAL trimming by whatever period we decide to make that control plane check-in.
+
+When discussing correctness, read "deletion" as meaning both the deletion of remote objects,
+and publishing updates to remote_consistent_lsn.
+
+### Index changes
+
+Since object keys now include a generation suffix, the index of these keys must also be updated.
+
+IndexPart currently stores keys and LSNs sufficient to reconstruct key names: this would be
+extended to store the generation numbers with which a particular object was written, to enable
+reconstructing the suffix as well.
+
+This will increase the size of the file, but only modestly: layers are already encoded as
+their string-ized form, so the overhead is about 20 bytes per layer. This will be less if/when
+the index storage format is migrated to a binary format from JSON.
+
+### Pageserver startup changes
+
+- The pageserver must obtain a generation number by some means (see Control Plane Changes below) before
+  doing any remote writes.
+- The pageserver _may_ also do some synchronization of its attachments to see which are still
+  valid, see "Synchronizing attachments on pageserver startup" in the optimizations section.
+
+### `pageserver::Tenant` startup changes
+
+#### Attachment
+
+We already reconcile with remote metadata during attachment. It is necessary to synchronize
+with remote metadata even if there is some local metadata that could be brought up to
+date by ingesting the WAL, because if another node updated
+the remote storage, it might have advanced the remote_consistent_lsn such that the safekeeper drops WAL
+content that would be needed for this node to recover. Reconcilation with remote metadata also acts as an optimization, to avoid redundantly
+re-ingesting WAL data that is already in S3.
```

Comment:
S3 uploads lag behind by 256MB or 10min. So, some re-ingesting of WAL during attach after node failure is inevitable anyways.